### ISSUE TITLE
There is a lot of UI lag and dropped UI interactions

### ISSUE THREAD

**Issue by foxmajik** [UnigramDev/Unigram#1272]
The UI has frequent jank and lag. It seems like you are giving priority to backend data handling and ignoring the importance of responsive UI.

Sometimes UI elements have to be clicked 2-3 times before they respond (ie, clicking to switch to a different chat often doesn't get a UI response) so it seems that you are dropping user inputs when the app is busy rather than buffering them.

UI responsiveness should take priority over data processing handling.

Unprocessed data should go into a buffer while you handle UI events, then after you handle the UI you should process the buffered data. Same for UI. Don't just ignore a user input when you're busy, put the input into a data buffer then handle it when you have resources.

User doesn't care if it takes longer for a message to appear but if you don't respond to their input or display stuff they are typing they want to stop using your app.

**Comment by MarcAnt01** [comment]
Unigram is fast and responsive here, what is your device hardware?

**Comment by FrayxRulez** [comment]
Any kind of input is handled by UWP (and then by Windows/Microsoft) and not by the app, that just consumes data provided by the system.

Your UX doesn't sound any normal tho, never experienced anything like this in 3 years developing and using this app (Even now that I'm forced to use it in a VM)

**Comment by AnggaSP** [comment]
Can confirm that UI has a lot of janks, especially with multiple account (I have 2 accounts, all have multiple groups and channels which usually have 1000+ chats in every group every day)

I can help profile it if you need the data.

**Comment by xperiandri** [comment]
For example, when you activate the search bar, then click on a quick suggestion. Then you have to wait around 2 seconds for a chat to appear.

**Comment by MarcAnt01** [comment]
> For example, when you activate the search bar, then click on a quick suggestion. Then you have to wait around 2 seconds for a chat to appear.

That sounds more pointless that the issue itself, since it depends on Telegram servers and using every client the delay is the same

**Comment by xperiandri** [comment]
That delay does not matter. UI must appear first (maybe with dummy placeholders until the server responds). But otherwise, you just don't understand what is going on

**Comment by FrayxRulez** [comment]
I think it's not a news that UWP lags like hell?

**Comment by MarcAnt01** [comment]
Well, I think that's a small detail that just 0,1% of users would notice, or at least looks in this way with my laptop (maybe in PCs with worse hw it'll be a lot worse)

**Comment by FrayxRulez** [comment]
If you click on a message sent by you in search results the app is going to lag due a bug in the Windows platform that Microsoft never fixed: https://github.com/microsoft/WindowsCompositionSamples/issues/329

**Comment by AnggaSP** [comment]
What about intermittent freeze when typing, navigating, or general use of the apps?
I sometimes encounter freeze for up to 2secs, UI is totally unresponsive, clicking anything won't do anything. I can still type tho (with it freeze but the text will eventually registered after it unfreezed)

Sometimes, text is doubled too, for example i type this and it also typed "nd it also typed" a chunk of it from micro freezes.

**Comment by xperiandri** [comment]
> (maybe in PCs with worse hw it'll be a lot worse)

Core i7 8650U, 16GB RAM, SSD NVMe

**Comment by MarcAnt01** [comment]
> > (maybe in PCs with worse hw it'll be a lot worse)
> 
> Core i7 8650U, 16GB RAM, SSD NVMe

Same hardware here, apart from the Ram

**Comment by xperiandri** [comment]
> I think it's not a news that UWP lags like hell?

I agree in general but not in my case. This case is a matter of algorithm and UX design

**Comment by FrayxRulez** [comment]
I think there’s a lot of confusion going on here as any action that require some loading always update the UI first and then loads the data.

**Comment by AnggaSP** [comment]
Can we go back to the original issue at hand?

**Comment by xperiandri** [comment]
> I think there’s a lot of confusion going on here as any action that require some loading always update the UI first and then loads the data.

But why then chat section of a window does not appear instantly when I click on a quick contact suggestion?

**Comment by FrayxRulez** [comment]
> But why then chat section of a window does not appear instantly when I click on a quick contact suggestion?
What is this?

**Comment by xperiandri** [comment]
![image](https://user-images.githubusercontent.com/2365592/71583913-09fd6680-2b19-11ea-90a1-b8e03f9e5bab.png)

**Comment by FrayxRulez** [comment]
Because UWP is really slow to initialize and load the UI

**Comment by xperiandri** [comment]
But lag only happens for quick contact suggestions. It does not happen for conversation opened from conversation list.

**Comment by FrayxRulez** [comment]
It doesn't happen for quick contact suggestions here

**Comment by MarcAnt01** [comment]
> It doesn't happen for quick contact suggestions here

I have understood that to reproduce you have to close the app and open it again, it seems to happen just the first time, once loaded the next times it won't happen

**Comment by FrayxRulez** [comment]
Then as I said it happens for the first chat you open, doesn't depend on the interaction source

**Comment by xperiandri** [comment]
No, it happens always. 1-2 seconds lag is rare while some comparably noticeable delay happens always

**Comment by hatzkel** [comment]
Here this just started happening recently as well for me. Here is an example:
![image](https://user-images.githubusercontent.com/48680234/71585126-837a6200-2ae3-11ea-9033-d446312aa5ae.png)

**Comment by MarcAnt01** [comment]
@hatzkel you are writing in the wrong ticket

**Comment by xperiandri** [comment]
@hatzkel, I had the same issue until I updated to 3.13.2647

**Comment by hatzkel** [comment]
I just updated and it started happening. :O 
I figured this was the right ticket since I was considering it "dropped UI interactions and or ui interaction lag" 🤷‍♀️

**Comment by MarcAnt01** [comment]
> I just updated and it started happening. :O
> I figured this was the right ticket since I was considering it "dropped UI interactions and or ui interaction lag" 🤷‍♀️

You have just posted it in the wrong ticket, creating confusion. Next time, kindly check the ticket description, from which you'll understand what it is about

**Comment by hatzkel** [comment]
Wow your English is so horrible that I don't even think you understand what I said. You don't see me trying to be smart in Italian maybe you should stick to coding and speaking Italian rather than handle support tickets written in English. I just quit using Unigram because you can kindly shove it up your ass and understand what it's about.

**Comment by FrayxRulez** [comment]
![image](https://user-images.githubusercontent.com/8152849/71588372-872cd980-2b21-11ea-9b3a-e974fad60985.png)

**Comment by AnggaSP** [comment]
https://streamable.com/0egbg

Hopefully this would illustrate the issue better

For context, my machine:
i3-4150
8GB DDR3-1600
256GB SATA SSD

**Comment by FrayxRulez** [comment]
You can try to disable visual effects from System settings.

**Comment by foxmajik** [comment]
I switched to Telegreat because it supports blocking users in group chats.  ¯\_(ツ)_/¯

**Comment by AnggaSP** [comment]
Same here, end up resorting to many instance telegram client and it's never been this fast
Sorry for OOT but I don't think anybody is looking at this issue.

**Comment by MarcAnt01** [comment]
> I switched to Telegreat because it supports blocking users in group chats. ¯_(ツ)_/¯

This is completely off topic and isn’t going to help at all here

**Comment by foxmajik** [comment]
> > I switched to Telegreat because it supports blocking users in group chats. ¯_(ツ)_/¯
> 
> This is completely off topic and isn’t going to help at all here

Pretty sure your comment isn't helpful either, and yet there it is.

Perhaps you should focus less on attacking users and more on fixing bugs?

**Comment by MarcAnt01** [comment]
1. I am not a developer
2. You don’t tell me what to do and I don’t owe you anything
3. How do you think that telling us you switched to another client is going to help somehow?

**Comment by foxmajik** [comment]
I switched to a different fork, so I'm done here.

**Comment by FrayxRulez** [comment]
> I switched to a different fork, so I'm done here.

Unigram isn’t a fork. And the fact that you moved to something else doesn’t mean that the problem is solved.

**Comment by Kenya-West** [comment]
- CPU: i5-3570K
- DDR: 32GB
- SSD: 256GB
- GPU: AMD Radeon R9 390X 8GB
- Unigram any kind of chats count: estimated ~2300

Can confirm, Unigram doesn't respond fast after sleep mode. After computer wakes up, Unigram can lag about 2-3 sec. and only then responds. I have close the app and then reopen it. Same behaviour on i7-6700K, 48GB, 2TB SSD, GTX 710.

**Comment by notthebee** [comment]
Same here. Unigram will become slow and laggy after some time. Text would take 5 seconds to actually appear after typing sometimes, mouse clicks won't register, etc.

**Comment by FrayxRulez** [comment]
IMHO most of this problems highly depend on UWP rather than on the app itself, I don't think that this framework was ever optimized for this kind of "real world scenario".
@notthebee do you frequently switch between multiple accounts?

**Comment by Lovemetal** [comment]
Same, it only happens when I switch between accounts (I have 3). Unigram will become laggy and unresponsive, I'll have to click UI elements multiple times before they respond. Also there will be double notification sound.

**Comment by FrayxRulez** [comment]
> Same, it only happens when I switch between accounts (I have 3). Unigram will become laggy and unresponsive, I'll have to click UI elements multiple times before they respond. Also there will be double notification sound.

Yes, this is understandable. When you switch between accounts a new view is created but due to a bug previous ones aren't ever garbage collected. This means that they will keep accumulating into app memory making it slower.
Honestly fixing this is absolutely non-trivial and I never managed to.

**Comment by Kenya-West** [comment]
@FrayxRulez thanks for explanation how it works! For me since 4.0 release there are no notable lags except only one: when I search for my own old (1+ year) message and click on it, Unigram and entire Windows 10 kinda freezes up for ~40 seconds - I can't click anything on any window, but music plays on, I can drag cursor anywhere. You can see this lag on a [video in OneDrive](https://1drv.ms/v/s!AjKPhFr4NBmvypU6F09U956sbWOwPg?e=HmX1C3) between 0:12 and through the end.

**Comment by FrayxRulez** [comment]
@Kenya-West this is a UWP bug that I reported so long ago that I can’t even remember where to find the ticket. It is caused by the list element tilt effect (the thing that makes element move in the “3D space” when you click on them) + the sent/read check mark, because of a internal UWP bug. 
Tilt effect for some reason can’t be disabled (chat list doesn’t have it because I spent really a lot of time writing a list component myself for this scope) and I guess that the only sane solution would be to remove the read/sent ticks from messages results list.

**Comment by Kenya-West** [comment]
> this is a UWP bug that I reported so long ago that I can’t even remember where to find the ticket

@FrayxRulez I found [a new official Microsoft repo](https://github.com/microsoft/windev) where you can possibly report the bug again. I guess this is a right channel

**Comment by lazerl0rd** [comment]
I'm not sure if this helps, but when transitioning from an administrator to standard account I started experiencing lags.

**Comment by lazerl0rd** [comment]
> I'm not sure if this helps, but when transitioning from an administrator to standard account I started experiencing lags.

UPDATE: It seems re-installing the application solves this issue but I'm able to reproduce it with another account. Does Unigram use administrator privileges for a certain file that it polls, maybe?

**Comment by VladWinner** [comment]
In addition to 1-2 second freezing and UI animations lags in the client, any video in built-in player is played with ≈ three times less FPS than it should be.
Both in 7.7 (MS Store) and 7.8 (Beta).
This is probably related to the same problem.

**Comment by FrayxRulez** [comment]
> In addition to 1-2 second freezing and UI animations lags in the client, any video in built-in player is played with ≈ three times less FPS than it should be.
> Both in 7.7 (MS Store) and 7.8 (Beta).
> This is probably related to the same problem.

I can only suggest you not to use Unigram then.

**Comment by VladWinner** [comment]
> I can only suggest you not to use Unigram then.

Personally, I would not like to abandon this full-featured client on my Windows 10 PC due to unspecified bugs leading to this kind of problems.

**Comment by FrayxRulez** [comment]
> > I can only suggest you not to use Unigram then.
> 
> Personally, I would not like to abandon this full-featured client on my Windows 10 PC due to unspecified bugs leading to this kind of problems.
> P.S. I only one Telegram account only on this client.

Sure thing, but then provide more context. If you launch the app and the video player - that’s a system component using hardware acceleration directly hosted within the app - plays videos three times slower than it should it’s not like I can’t do magic.

**Comment by VladWinner** [comment]
> Sure thing, but then provide more context.

Now I noticed that any video (when viewed in Unigram) plays smoothly (as it should) in the pop-up mode, but in normal playback (right after clicking on the video icon) FPS is reduced, which does not affect the speed of the video and audio stream. With "Enable full-screen gallery" option enabled problem still exist.
It is clearly not a low-performance computer issue (in official Telegram client, in system UWP-based player and in Win32 videoplayers like MPC-HC, mpv or VLC there is also no problem).
What kind of context or information can I provide to solve the bug?

**Comment by FrayxRulez** [comment]
What’s the difference between normal mode and pop up mode?

**Comment by VladWinner** [comment]
> What’s the difference between normal mode and pop up mode?

Reducing the number of frames per second (FPS) in video playback. I recorded a [small video clip](https://www.youtube.com/watch?v=nmzum0kE8g0) to show the difference.

**Comment by VladWinner** [comment]
> Reducing the number of frames per second (FPS) in video playback (normal mode). I recorded a [small video clip](https://www.youtube.com/watch?v=nmzum0kE8g0) to show the difference.

Maybe it's weird bug related to **FPS limit**er for stickers or VideoAnimation code or whatever?

\Unigram.Native\VideoAnimation.cpp
![image](https://user-images.githubusercontent.com/13555921/123010458-51d81380-d3c7-11eb-8943-2457cb7e3151.png)

\Controls\DiceView.cs
![image](https://user-images.githubusercontent.com/13555921/123010105-9adb9800-d3c6-11eb-991e-58b468d9a9c4.png)

\Controls\LottieView.cs
![image](https://user-images.githubusercontent.com/13555921/123010157-b2b31c00-d3c6-11eb-969a-39f393fa8f56.png)

I checked Unigram 7.8 (6559) — the bug still remains.

**Comment by FrayxRulez** [comment]
None of those widgets are related to the gallery. Furthermore the player in the PiP window is the same from the main window gallery (that’s it, it gets removed from the view and added to the new one) - that’s a 100% system component and there’s even no way to control FPS, unfortunately.

**Comment by FrayxRulez** [comment]
> On top of that, when Unigram is in foreground and is minimized, or any other app is opened, they drop frames as well.

When Unigram is minimized the process is suspended by the operating system and no code is executed, you can check that from your Task Manager.

**Comment by pratyakshm** [comment]
> When Unigram is minimized the process is suspended by the operating system and no code is executed, you can check that from your Task Manager.

Sorry that was a typo from my side. Edited and fixed the comment now.

**Comment by FrayxRulez** [comment]
> Sorry that was a typo from my side. Edited and fixed the comment now.

Maybe you should monitor your Task Manager anyway? UWP apps are sandboxed and aren’t supposed to affect other softwares - apart from that Unigram doesn’t do anything that heavy almost ever and average CPU usage is 1-4%.

**Comment by VladWinner** [comment]
> None of those widgets are related to the gallery. Furthermore the player in the PiP window is the same from the main window gallery (that’s it, it gets removed from the view and added to the new one) - that’s a 100% system component and there’s even no way to control FPS, unfortunately.

I just discovered a partial workaround (if you have Nvidia GPU).
1. Go to NVIDIA Control Panel → 3D Settings → Manager 3D settings → Program Settings:
![image](https://user-images.githubusercontent.com/13555921/127748220-8e4da5fc-ef5f-4400-9d83-a8e57a1c9db8.png)
2. Click 'Add' button and select Unigram app (it will be in recents list)
![image](https://user-images.githubusercontent.com/13555921/127748405-41b73747-7b1b-4af3-88e4-f35025a48291.png)
3. Then set the 'Vertical sync’ to 'Use the 3D application setting'.
4. (Optional for G-Sync monitors) You may have to change 'Monitor technology' parameter towards the maximum number of frames (I don't have G-Sync to name the exact option)
5. Click on '1. Select a program to customize' list and repeat steps 3-4 for following apps:
- Metro Video Player (Movies & TV)
- Windows Photo Viewer (Microsoft Photos)

But in fact, this solution returns smoothness only when video file are played in full-screen mode (after mouse click on video thumbnail/preview) while video messages and video file thumbnails/previews remain with the same reduced frame rate.

**Comment by FrayxRulez** [comment]
@adivvkvns There’s no way to fix the issue as Microsoft Windows and UWP are not open source and consequently performances can’t be optimized there. Everything possible performance-wise is already done in the client code. At the current state there are no plans to migrate to another framework, as there’s no other desktop framework as complete and modern as UWP. The only hope is that WinUI 3 will be production ready and open source soon, and that Microsoft will invest some energy in improving performances there.

**Comment by adivvkvns** [comment]
> @adivvkvns There’s no way to fix the issue as Microsoft Windows and UWP are not open source and consequently performances can’t be optimized there. Everything possible performance-wise is already done in the client code. At the current state there are no plans to migrate to another framework, as there’s no other desktop framework as complete and modern as UWP. The only hope is that WinUI 3 will be production ready and open source soon, and that Microsoft will invest some energy in improving performances there.

Thanks for the answer. A little sorry.
But I have not seen a better messenger than this one😁

**Comment by luccascomvoce** [comment]
I don't often use Unigram for calls and voice/video chats because screen sharing is absent or problematic. But today, trying to join voice chats through Unigram, I experienced a lot of delay in loading the interface and connecting to the server.

https://user-images.githubusercontent.com/65538684/188293578-6355b2d3-3253-4d14-ac26-4ef0a3123824.mp4

Unigram 8.9(7616) x64
Windows 10 Home 21H2
Ryzen 7 4800H, 8GB RAM 3200 GHz, SSD NVME 512GB

**Comment by ti-bone** [comment]
@luccascomvoce Does it happens to you in Telegram Desktop?

**Comment by luccascomvoce** [comment]
> 

No, anyway. My experience with Telegram Desktop (including the tdesktop beta) is absolutely stable without any inconvenience.

**Comment by luccascomvoce** [comment]
@ti-bone  No, anyway. My experience with Telegram Desktop (including the tdesktop beta) is absolutely stable without any inconvenience.

**Comment by FrayxRulez** [comment]
@luccascomvoce can you create a dedicated ticket about calls issue?

**Comment by MarcAnt01** [comment]
@FrayxRulez they already opened it (https://github.com/UnigramDev/Unigram/issues/2845)

**Comment by shanoaice** [comment]
To me this bug seems to happen when I scroll very far from the bottom of the chat, loading a lot of chat history. Maybe this could be an issue relating to deleting elements some distance outside of the viewport?

**Comment by MinhThienDX** [comment]
> > None of those widgets are related to the gallery. Furthermore the player in the PiP window is the same from the main window gallery (that’s it, it gets removed from the view and added to the new one) - that’s a 100% system component and there’s even no way to control FPS, unfortunately.
> 
> I just discovered a partial workaround (if you have Nvidia GPU).
> 
>     1. Go to NVIDIA Control Panel → 3D Settings → Manager 3D settings → Program Settings:
>        ![image](https://user-images.githubusercontent.com/13555921/127748220-8e4da5fc-ef5f-4400-9d83-a8e57a1c9db8.png)
> 
>     2. Click 'Add' button and select Unigram app (it will be in recents list)
>        ![image](https://user-images.githubusercontent.com/13555921/127748405-41b73747-7b1b-4af3-88e4-f35025a48291.png)
> 
>     3. Then set the 'Vertical sync’ to 'Use the 3D application setting'.
> 
>     4. (Optional for G-Sync monitors) You may have to change 'Monitor technology' parameter towards the maximum number of frames (I don't have G-Sync to name the exact option)
> 
>     5. Click on '1. Select a program to customize' list and repeat steps 3-4 for following apps:
> 
> 
>     * Metro Video Player (Movies & TV)
> 
>     * Windows Photo Viewer (Microsoft Photos)
> 
> 
> P.S. Selecting any other parameter except 'Use the 3D application setting' and 'Use global setting' causes me to flicker the entire Unigram window and a strange glow effect on him.
> 
> But in fact, this solution returns smoothness only when video files are played in full-screen mode (after mouse click on video thumbnail/preview) while video messages and video file thumbnails/previews remain with the same reduced frame rate.

Turning off FreeSync for this app works too

**Comment by AqlaSolutions** [comment]
I see lots of lags too after switching between multiple accounts and chats while participating in >100 active groups. According to the memory profiler garbage collection frequently freezes UI thread for several seconds. I performed lots of optimizations to reduce allocations but the result is still far from good. It looks like these GC pauses are caused by UWP, not just TDLib spamming updates. If anyone could point me where to look at I would investigate it further.

**Comment by FrayxRulez** [comment]
> If anyone could point me where to look at I would investigate it further.

UWP and consequently Unigram suffer from memory leaks due to its faulty architecture and the fact of having a managed app running on top of a native framework. 
Every now and then I try to do some profiling, although it usually is a waste of time as C# profilers can't see the memory allocated by the framework, so it's quite hard to investigate into this (ie even if the app is using 1 GB of memory .NET profilers will always show around ~10 MB of used RAM).
The problem isn't indeed TDLib, but the fact that some UI components are never freed from memory and they accumulate over and over, adding a lot of pressure and causing more and more frequent GC collections.
As an example, when you switch between accounts the whole UI gets recreated, and there are definitely some scenarios where the previous UI gets completely leaked. Unfortunately, again, profilers (the only one that seems to work decently with UWP is dotMemory btw, as the one built-in VS tends to crash or stop working completely) don't help much here, because they're unable to find what causes the leak due to the fact that all framework objects have native handles and so everything shows up as "self/cyclic reference".
Another problem is that execution mode changes completely between debug and release builds, as release uses .NET Native and some behaviors (and issue) may exist only in one configuration and not in the other one.

**Comment by AqlaSolutions** [comment]
@FrayxRulez , thank you for your answer, unfortunately long time passed since I stopped development of my fork but hopely someone else will find this helpful.

I want to point out that long GC pauses can't be caused by unmanaged memory. GC time is only spent on walking through managed objects. Memory leaks are not the main problem here. Many instances of TDLib cause very high memory traffic with all those short-living events objects with chat updates. I think that those events also cause your code to recreate lots of managed UI objects (which may point to unmanaged but I'm talking about managed memory traffic now). Also it's possible that your code still tries to update UI for non-foreground accounts.

**Comment by FrayxRulez** [comment]
Many of the unmanaged objects have pinned handles in the managed heap. This includes by example all event handlers registration. UWP also does a lot of caching for types initialization, still in the managed heap. All this stuff is going to build up and slow down GC collection for sure, and they as well increase the pressure: while the managed handle is just a few bytes, they still report that these objects occupy space in the native heap as well.
Sure thing, TDLib may cause some traffic with all the updates, but for the vast majority of users this shouldn't have any real impact, as most reports point out that freezes occur on interaction (so, to say, opening a new screen generates enough traffic to trigger a collection).
The code doesn't try to update UI for non-foreground accounts.

**Comment by AqlaSolutions** [comment]
I see your point. However, those objects that you named are usually long-living which means that they end up in 2nd GC generation so their collection is performed in a background thread without affecting UI performance. But those stutters for 1-3-5 seconds that I notice while typing a message can only be caused by stop-the-world 0 and 1 generation collection.

**Comment by MinhThienDX** [comment]
For me, this issue is fixed.
Not sure how.

**Comment by Looz-Ashae** [comment]
Still it's an issue. It was very interesting to know, that it's not some electron, but a native c#. It should be fast and responsive AF. Yet here we are.

**Comment by FrayxRulez** [comment]
@Looz-Ashae UWP apps written in C# can’t really be that responsive because UWP itself is written in C++ and exposed as COM, that is known to be quite slow especially in interop contexts. Secondarily, still due to the C#/C++ mix, GC can’t really work well because the garbage collector is not really aware of what’s going on and after a while it’s just going to trigger a lot of garbage collections slowing down everything. Third, UWP is known quite a lot of memory leaks on its own, that are just going to make the previous problem worse. And last but not least the abandonment of the framework and the complete lack of proper tooling (also because of the C++/C# hybrid) makes it hard if not impossible to do any kind of memory optimization on top of all the problems previously mentioned.

**Comment by Kenya-West** [comment]
Also worth mentioning that Unigram utilizes nearly ~~146%~~ 100% of UWP capabilities and already met UWP boundaries years ago. UWP, technically, is just not the scale for such a good complex app.

**Comment by FrayxRulez** [comment]
@Kenya-West There’s no desktop framework (apart from maybe SwiftUI/Cocoa on macOS) to develop such an app properly, at the moment, the best option are probably web technologies as they’re crazily optimized nowadays.

**Comment by xperiandri** [comment]
Works perfect now. I don't see any lags anymore
